`RowVector` integration for PyVelox

vibhatha opened the pull request:
This PR includes `RowVector` creation in PyVelox. Also this is an important feature for the continuation of https://github.com/facebookincubator/velox/pull/5023.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +519,48 @@ static void addVectorBindings(
+          childTypes[i] = values[i]->type();
+        }
+        auto rowType = ROW(std::move(names), std::move(childTypes));
+        const size_t vectorSize = values.empty() ? 0 : values.front()->size();
```
Its possible to have legal vectors with child size < row size , if the rest of the row vectors consists of trailing nulls.

kgpai replied:
Example consider a Row with one child of size 10 and parent of size 20. Here this is legal if the parent has nulls from index 10 to 19.

vibhatha replied:
I understand, then how should we set the `vectorSize` would it be accurate to assume it as the size of child elements?

kgpai replied:
The vector size will be size of row (thus 20) , the child sizes can be different though.

kgpai replied:
In this case for simplicity now we can enforce ensuring that all children and parent are of same size.

vibhatha replied:
I updated the code based on this discussion.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +519,48 @@ static void addVectorBindings(
+        auto rowVector = std::make_shared<RowVector>(
+            PyVeloxContext::getSingletonInstance().pool(),
+            rowType,
+            BufferPtr(nullptr),
```
Its better to set the nulls buffer upfront rather than using setNulls later.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +519,48 @@ static void addVectorBindings(
+      "row_vector",
+      [](std::vector<std::string>& names,
+         std::vector<VectorPtr>& values,
+         std::function<bool(vector_size_t /*row*/)> isNullAt = nullptr) {
```
Not sure that this is very useful to have at as a function (it makes sense when writing test cases).

vibhatha replied:
@kgpai would you rather suggest an input vector with flags for null in boolean and then use that to set the nulls?

vibhatha replied:
And it would resolve https://github.com/facebookincubator/velox/pull/6533#discussion_r1323668759 as well?

kgpai replied:
Probably better to have an input vector, as we can use list comprehensions etc in python to generate it.

vibhatha replied:
Sure will do that.

vibhatha commented:
@kgpai I updated the PR and left some comments too.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -27,12 +28,16 @@
 #include "context.h"
 
+/// TODO: remove
+#include <iostream>
```
Doesnt LOG(INFO) work in pyvelox ? Anyways please remove.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +532,56 @@ static void addVectorBindings(
+      },
+      py::arg("names"),
+      py::arg("values"),
+      py::arg("nullability") = std::nullopt);
```
nullability is not a name that is common in velox, maybe just nulls ?

vibhatha commented:
Thanks for the review @kgpai. I will address these.

vibhatha commented on pyvelox/pyvelox.h:
```diff
@@ -109,6 +114,29 @@ inline velox::variant pyToVariant(const py::handle& obj, const Type& dtype) {
+inline std::string rowVectorToString(const RowVectorPtr& v) {
+  if (v->childrenSize() > 0) {
+    auto child = v->childAt(0);
+    size_t limit = child->size();
```
I updated the code. And I see your point.

vibhatha commented:
@kgpai ready for review. Appreciate your feedback.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -109,6 +111,27 @@ inline velox::variant pyToVariant(const py::handle& obj, const Type& dtype) {
+}
+
+inline VectorPtr getVectorFromRowVectorPtr(
+    const RowVectorPtr& v,
```
nit: having short names for vector (like v) goes against convention in this file and in velox.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -109,6 +111,27 @@ inline velox::variant pyToVariant(const py::handle& obj, const Type& dtype) {
+
+inline std::string rowVectorToString(const RowVectorPtr& v) {
+  if (v->childrenSize() > 0) {
+    return v->toString(0, v->size());
```
why not just return v->toString(0, v->size()) ? It should work even if there are no children..

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+        std::vector<std::shared_ptr<const Type>> childTypes;
+        childTypes.resize(values.size());
+        for (int i = 0; i < values.size(); i++) {
+          childTypes[i] = values[i]->type();
```
nit: use push_back ?

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+        }
+        auto rowType = ROW(std::move(names), std::move(childTypes));
+        const size_t vectorSize = values.size();
+        BufferPtr nullability_buffer = nullptr;
```
nit: lets not mix python style names with velox style (nullability_list, rowType ) etc.. Would be nice to stick to one convention.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+  m.def(
+      "row_vector",
+      [](std::vector<std::string>& names,
+         std::vector<VectorPtr>& values,
```
nit: values seems like bad name, lets name it children ?
Should this be a const reference or is the idea that these get moved into the rowvector ?

vibhatha replied:
I think it would be better to move? WDYT?

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+          childTypes[i] = values[i]->type();
+        }
+        auto rowType = ROW(std::move(names), std::move(childTypes));
+        const size_t vectorSize = values.size();
```
vectorSize != number of children .

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+            PyVeloxContext::getSingletonInstance().pool(),
+            rowType,
+            nullability_buffer,
+            vectorSize,
```
Ideally we should take the size of the vector from the user (maybe optional argument) , else compute it as smallest child size if the child sizes are different ? But I feel  we should be explicit - I prefer throwing an error if children are different sizes.

vibhatha replied:
Good catch. I will add the changes.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+      "row_vector",
+      [](std::vector<std::string>& names,
+         std::vector<VectorPtr>& values,
+         const std::optional<py::list>& nullability_list) {
```
Is it possible to pass a generator / lambda as the nullability_list, i cant imagine many people will like to put true, false, true etc..
maybe even option for a dict of rowNumber:true/false , with no value being understood as rowNumber:true ?

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +527,56 @@ static void addVectorBindings(
+        if (nullability_list.has_value()) {
+          auto nullability_values = nullability_list.value();
+          nullability_buffer = AlignedBuffer::allocate<bool>(
+              nullability_values.size(),
```
nullability size can then be different than vector size, they should all be one size, irrespective of number of elements in nullability_list...

vibhatha commented:
I will address these reviews @kgpai.

jorisvandenbossche commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,64 @@ static void addVectorBindings(
+
+  py::class_<RowVector, BaseVector, RowVectorPtr>(
+      m, "RowVector", py::module_local(asModuleLocalDefinitions))
+      .def("__len__", &RowVector::childrenSize)
```
FWIW, I think it will be confusing that `len(..)` for RowVector works differently than for other vector types. My understanding is that for BaseVector, this is the number of values (i.e. the logical length of the vector), while here you make it the number child vectors. 

Another option is to add a specific attribute (like `num_children`) for this, and keep `__len__` consistent with the base vector.

vibhatha replied:
I see your point. We could change it quite easily. @kgpai any thoughts. I am +1 with this change.

jorisvandenbossche commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,64 @@ static void addVectorBindings(
+      .def("__len__", &RowVector::childrenSize)
+      .def("__str__", [](RowVectorPtr& v) { return rowVectorToString(v); })
+      .def("may_have_nulls", &RowVector::mayHaveNulls)
+      .def("__getitem__", [](RowVectorPtr& v, vector_size_t idx) {
+        return getVectorFromRowVectorPtr(v, idx);
```
Similarly as my comment above, I _think_ for BaseVector doing `vec[10]` gives you the 10th value, while here this would be the 10th child?

vibhatha replied:
Yes that is correct.

vibhatha replied:
Shouldn't it be the case? We want to access an item from the RowVector it should be a BaseVector? Is this logic wrong?

vibhatha replied:
Ignore the previous comment. I understand your point.

kgpai replied:
Lets return a dictionary here , with child names as keys and value at idx for child as value.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,67 @@ static void addVectorBindings(
+        if (nullabilityDict.has_value()) {
+          auto nullabilityValues = nullabilityDict.value();
+          nullabilityBuffer = AlignedBuffer::allocate<bool>(
+              values.size(),
```
should be vectorSize I think.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,67 @@ static void addVectorBindings(
+                  "nullability must be a dictionary, rowId in int and nullability in boolean");
+            }
+            int rowId = py::cast<int>(row);
+            bool nullabilityVal = py::cast<bool>(nullability);
```
check if rowId < vectorSize.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,67 @@ static void addVectorBindings(
+      .def("may_have_nulls", &RowVector::mayHaveNulls)
+      .def("__getitem__", [](RowVectorPtr& v, vector_size_t idx) {
+        return getVectorFromRowVectorPtr(v, idx);
+      });
```
Also lets add `childAt` method to return vector at idx'i' . 
Should we also consider rowvector like dictionary ; so if I have row('a':<somevector>, 'b':<someothervector>), then row['a'] gives me <somevector>.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,67 @@ static void addVectorBindings(
+        std::vector<std::shared_ptr<const Type>> childTypes;
+        childTypes.reserve(values.size());
+        size_t vectorSize = values.size() > 0 ? values[0]->size() : 0;
+        for (int i = 0; i < values.size(); i++) {
```
lets make sure all child vectors have same size.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +523,67 @@ static void addVectorBindings(
+      m, "RowVector", py::module_local(asModuleLocalDefinitions))
+      .def("__len__", &RowVector::childrenSize)
+      .def("__str__", [](RowVectorPtr& v) { return rowVectorToString(v); })
+      .def("may_have_nulls", &RowVector::mayHaveNulls)
```
should we also have a bind to rowVector:;resize() ?

vibhatha commented:
@kgpai I addressed the reviews to get the core functionality which could support us in the [PyVelox Substrait PR](https://github.com/facebookincubator/velox/pull/5023). 
I created an issue to carry out the missing pieces [here](https://github.com/facebookincubator/velox/issues/7359).

vibhatha commented:
@kgpai there is a vector save function which is important for the Substrait PR, would it be okay to add it to this PR or do you prefer a separate PR for this?

kgpai commented:
@vibhatha Lets create separate PR for save functionality.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +532,75 @@ static void addVectorBindings(
+        childTypes.reserve(children.size());
+        size_t vectorSize = children.size() > 0 ? children[0]->size() : 0;
+        for (int i = 0; i < children.size(); i++) {
+          // choose child with smallest size to calculate the vector size
```
This comment is misleading with the error message.

vibhatha replied:
Yes, we changed the logic. I will update.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +532,75 @@ static void addVectorBindings(
+        }
+        std::vector<std::shared_ptr<const Type>> childTypes;
+        childTypes.reserve(children.size());
+        size_t vectorSize = children.size() > 0 ? children[0]->size() : 0;
```
In our case I think we should just assert children is not empty and then set vectorSize = children[0]->size().

kgpai commented on pyvelox/test/test_vector.py:
```diff
@@ -351,3 +351,52 @@ def test_roundtrip_conversion(self):
+        )
+        assert expected_str == rw_str
+
+    def test_row_vector_comparison(self):
```
Can you add a test with different types (one stringview, one double, another ROW ) ?

vibhatha replied:
Sure, I will add.

kgpai reviewed (commented):
Looks good ; just some minor nits and one additional test.

kgpai commented on pyvelox/pyvelox.h:
```diff
@@ -504,6 +534,75 @@ static void addVectorBindings(
+        size_t vectorSize = children[0]->size();
+        for (int i = 0; i < children.size(); i++) {
+          if (i > 0 && children[i]->size() != vectorSize) {
+            PyErr_SetString(PyExc_ValueError, "Each child must have same");
```
"each child must have same size"

vibhatha commented:
@kgpai @jorisvandenbossche I updated the PR. 
Though there are two CIs failing and the same one is failing locally too, but I am not exactly sure how these are related.

kgpai commented:
I will rebase aginst main, the failures arent related to your changes.
